errata-query-4: CONCAT and COALESCE

afs opened the pull request:
This addresses [sparql-errata#errata-query-4](https://www.w3.org/2013/sparql-errata#errata-query-4).

This is part of #8.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/sparql-query/pull/70.html" title="Last updated on May 14, 2023, 7:18 PM UTC (e5e7387)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/sparql-query/70/b279bac...e5e7387.html" title="Last updated on May 14, 2023, 7:18 PM UTC (e5e7387)">Diff</a>

hartig commented on spec/index.html:
```diff
@@ -6814,12 +6818,16 @@ <h5>ENCODE_FOR_URI</h5>
           <section id="func-concat">
             <h5>CONCAT</h5>
             <pre class="prototype nohighlight"><span class="return">string literal</span>  <span class="operator">CONCAT</span>(<span class="type">string literal</span> <span>ltrl<sub>1</sub></span> ... <span class="type">string literal</span> <span>ltrl<sub>n</sub></span>)</pre>
```
```suggestion
            <pre class="prototype nohighlight"><span class="return">string literal</span>  <span class="operator">CONCAT</span>(<span class="type">string literal</span> <span>ltrl<sub>1</sub></span>, ..., <span class="type">string literal</span> <span>ltrl<sub>n</sub></span>)</pre>
```

hartig commented on spec/index.html:
```diff
@@ -6814,12 +6818,16 @@ <h5>ENCODE_FOR_URI</h5>
-              forms of its inputs. If all input literals are literals with identical language
-              tags, then the returned literal is a literal with the same language tag, in all
-              other cases, the returned literal is a literal with datatype <code>xsd:string</code>.</p>
+            <p>The <code>CONCAT</code> function returns a string literal following the 
+              <a href="#string-literal-return-type">string literal return type</a> rule.
+              The <a data-cite="RDF12-CONCEPTS#dfn-lexical-form">lexical form</a> of the
+              returned string literal is obtained by concatenating the lexical
```
I notice that the cross-referenced section (i.e., Section 17.4.3.1.3 String Literal Return Type) does not actually state any rule for CONCAT. Instead, it only says that "The function CONCAT returns a string literal based on the details of all its arguments." So, when coming to that section, the reader is essentially referred back to this section here. This circular cross-referencing is somewhat contradictory. Therefore, I propose to simply remove the reference from this section here.
```suggestion
            <p>The <code>CONCAT</code> function returns a string literal such that
              the <a data-cite="RDF12-CONCEPTS#dfn-lexical-form">lexical form</a> of
              this literal is obtained by concatenating the lexical
```

hartig commented on spec/index.html:
```diff
@@ -6814,12 +6818,16 @@ <h5>ENCODE_FOR_URI</h5>
+              If zero arguments are given, the result is an empty string literal without language tag.
+              If all input literals are literals with identical language tag,
+              then the returned literal is a literal with that language tag.
+              Otherwise, the returned literal is a literal with datatype <code>xsd:string</code>.
```
```suggestion
              Otherwise, the returned literal is a literal with datatype <code>xsd:string</code>
              (and no language tag).
```

rubensworks commented on spec/index.html:
```diff
@@ -6813,13 +6817,17 @@ <h5>ENCODE_FOR_URI</h5>
           <section id="func-concat">
             <h5>CONCAT</h5>
-            <pre class="prototype nohighlight"><span class="return">string literal</span>  <span class="operator">CONCAT</span>(<span class="type">string literal</span> <span>ltrl<sub>1</sub></span> ... <span class="type">string literal</span> <span>ltrl<sub>n</sub></span>)</pre>
-            <p>The <code>CONCAT</code> function corresponds to the XPath <a data-cite="XPATH-FUNCTIONS#func-concat">fn:concat</a> function. The function accepts string
```
As this doesn't rely on `fn:concat` anymore, we may want to add some details regarding the handling of unicode, which were implicit due to the reliance on `fn:concat`.

Specifically, I'm referring to the notes in https://www.w3.org/TR/xpath-functions/#func-concat

afs commented:
Added an example for `COALESCE()` (no args). Split paragraph in `COALESCE` to be similar in `CONCAT` layout.

@rubensworks - `CONCAT` rewritten to mention `fn:concat` for 2+ arguments.

afs commented:
@hartig FYI - removed the indices `1` and `n` on the function signature to lessen implication that zero or one argument isn't covered.

See also: https://github.com/w3c/rdf-tests/pull/100
